### PR TITLE
Use Cases - Regression

### PR DESCRIPTION
**Pull request by pandeva**
My first version of the regression use case. I would be glad about any feedback for improving it. (For example, I couldn't find a more elegant way to extract the tuned parameters.)

### CODE REVIEW

File: src/usecase_regression.Rmd [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+* tune the model ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/tune/index.html)),
+* conduct a benchmark experiment ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/benchmark_experiments/index.html)) and
+* evaluate the performance of the model ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/performance/index.html)).
+
+First, we have a look at a part of the data.
+```{r}
+data(BostonHousing, package = "mlbench")
+head(BostonHousing)
+```
+We have 14 numeric variables and as before ``medv`` is chosen to be the response variable for the regression model.
```
Comment (gcskoenig) [inline src/usecase_regression.Rmd]: Consider starting a new sentence here, imo more readable:
The dataset contains 14 numeric variables. As before, ...

File: src/usecase_regression.Rmd [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+### Performance
+
+Now we want to evaluate the results.
+```{r}
+head(getBMRPerformances(bmr, as.df = TRUE))
+
+```
+
+A closer look at the first boxplot reveals that random forest outperforms the other learners for this specific task. However, the SVM model using the measure ``mae`` seems to perform better than the other models. Moreover, in both cases the benchmark experiment for the linear regression and lasso regression models yield similar but poor results.
```
Comment (gcskoenig) [inline src/usecase_regression.Rmd]: Not sure if "Moreover" is suitable in this context. Something like this: (?)
"In both cases, the benchmark experiments for linear and lasso regression yield similar but poor results."

**Review by gcskoenig — COMMENTED** [review]
Great use case, added minor suggestions.

File: src/usecase_regression.Rmd (hunk 2 above)
Comment (pandeva) [reply src/usecase_regression.Rmd]: Thanks! :)

File: src/usecase_regression.Rmd [hunk 3]
```diff
@@ -11,7 +11,7 @@ First, we have a look at a part of the data.
 data(BostonHousing, package = "mlbench")
 head(BostonHousing)
 ```
-We have 14 numeric variables and as before ``medv`` is chosen to be the response variable for the regression model. 
+We have 14 variables. As before, ``medv`` is chosen to be the response variable.
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: * The use case should be a stand-alone example. If the "As before" refers to another page in the tutorial I would remove it.
* I guess that most people will know this data set. But could you please explain shortly (just a sentence) what the Boston Housing is about and what the response variable `medv` means?

File: src/usecase_regression.Rmd [hunk 4]
```diff
@@ -0,0 +1,107 @@
+# Regression
+For the regression use case we use the well-known [Boston Housing](https://www.rdocumentation.org/packages/mlbench/versions/0.4-1/topics/BostonHousing) dataset. The road map is as follows:
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: * Please link to `[Boston Housing](&mlbench::BostonHousing)`.
* Please add here (or below) a sentence explaining what the data set is about and what you want to predict.

File: src/usecase_regression.Rmd [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+* define the learning task  ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/task/index.html)),
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: * Please link to `[here](task.md)`.
* Same procedure for the links below.

File: src/usecase_regression.Rmd [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+* tune the model ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/tune/index.html)),
+* conduct a benchmark experiment ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/benchmark_experiments/index.html)) and
+* evaluate the performance of the model ([here](http://mlr-org.github.io/mlr-tutorial/devel/html/performance/index.html)).
+
+First, we have a look at a part of the data.
+```{r}
+data(BostonHousing, package = "mlbench")
+head(BostonHousing)
+```
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: * I think a `summary` would be nice, too.
* I would change the sentence in l.9 to "First, let's have a look at the data."

File: src/usecase_regression.Rmd [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+We have 14 variables. As before, ``medv`` is chosen to be the response variable. 
+
+### Define a task
+Let's define the regression [task](task.md).
+
+```{r}
+## Make a task
+regr.task = makeRegrTask(data = BostonHousing, target = "medv")
+regr.task$task.desc$n.feat
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: I would remove l. 22 and just print the `regr.task`. The overview of feature types is included there.

File: src/usecase_regression.Rmd [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+```
+With the command ``regr.task$task.desc$n.feat`` the type of all variables in the data can be accessed. 
+
+### Tuning
+By calling ``listLearners("regr")`` we can see which learners are available for the regression task. We picked the classical linear regression model (``regr.lm``), the [Lasso](https://www.rdocumentation.org/packages/penalized/versions/0.9-50) regression model (``regr.glmnet``), [SVM](https://www.rdocumentation.org/packages/kernlab/versions/0.9-25) with a radial basis kernel (``regr.ksvm``) and random forest from the [ranger](https://www.rdocumentation.org/packages/ranger/versions/0.6.0) package (``regr.ranger``), which we would like to compare. Before setting up a [benchmark experiment](benchmark_experiments.md) some of the models' hyperparameters can be tuned. The [%mlr] package provides powerful tuning algorithms. See [Tuning](tuning.md) and [Advanced Tuning](advanced_tune.md) for more details.
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: I would not link to the packages, but to the manual pages for the learners you use, for example `[Lasso](&glmnet::glmnet)`.

File: src/usecase_regression.Rmd [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+We start with specifying a search space for each parameter. Note that for all learners the hyperparameters, which will be tuned, should be positive. With [&makeTuneControlRandom] we set the tuning method to be random search. Afterwards we take $10$-fold cross validation as our [resampling strategy](resample.md) and root mean squared error (``rmse``) as optimization criterion.
+
+```{r}
+set.seed(1234)
+
+## Define the learners for the tuning
+base.learners = list(
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: You could also use `makeLearners` here.

File: src/usecase_regression.Rmd [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+By calling ``listLearners("regr")`` we can see which learners are available for the regression task. We picked the classical linear regression model (``regr.lm``), the [Lasso](https://www.rdocumentation.org/packages/penalized/versions/0.9-50) regression model (``regr.glmnet``), [SVM](https://www.rdocumentation.org/packages/kernlab/versions/0.9-25) with a radial basis kernel (``regr.ksvm``) and random forest from the [ranger](https://www.rdocumentation.org/packages/ranger/versions/0.6.0) package (``regr.ranger``), which we would like to compare. Before setting up a [benchmark experiment](benchmark_experiments.md) some of the models' hyperparameters can be tuned. The [%mlr] package provides powerful tuning algorithms. See [Tuning](tuning.md) and [Advanced Tuning](advanced_tune.md) for more details.
+
+We start with specifying a search space for each parameter. Note that for all learners the hyperparameters, which will be tuned, should be positive. With [&makeTuneControlRandom] we set the tuning method to be random search. Afterwards we take $10$-fold cross validation as our [resampling strategy](resample.md) and root mean squared error (``rmse``) as optimization criterion.
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: Could you please state which hyperparameters you are going to tune?

File: src/usecase_regression.Rmd (hunk 8 above)
Comment (schiffner) [reply src/usecase_regression.Rmd]: Could you please update this paragraph? It seems that in your code below, you are using slightly different learners than stated above.

File: src/usecase_regression.Rmd [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+num.trees = list("num.trees" = res.rf$regr.ranger.num.trees[which.min(res.rf[, 2])])
+```
+Finally, we extracted the optimal parameter value for each learner. Now we are ready to proceed with the benchmark experiment!
+
+### Benchmark Experiment
+In order to conduct a [benchmark experiment](benchmark_experiments.md), it is necessary to choose an evaluation method. We will use the resampling strategy from the previous section and two performance measures - root mean squared error (``rmse``) and mean of absolute error (``mae``).
+
+```{r}
+## Four learners to be compared
+lrns = list(makeLearner("regr.lm"),
```
Comment (schiffner) [inline src/usecase_regression.Rmd]: Could you please indent your code by 2 spaces?

**Review by schiffner — CHANGES_REQUESTED** [review]
Hi, thx very much.

I left some comments regarding technical stuff.

Generally and with regard to your question: The model multiplexer optimizes over several learners and their parameters. So in contrast to other tuning methods it does't just decide on optimal parameters but also on the best out of several candidate learners. What you are doing in your use case, i.e. extracting optimal parameter values for *each* of the candidate learners is against the spirit of the model multiplexer.
My recommendation would be to change your example as follows:
* Make tuning wrappers for all candidate learners,
* Pass these tuning wrappers to benchmark.

**Comment by berndbischl** [comment]
> The model multiplexer optimizes over several learners and their parameters.

please dont use the modelmultiplexer in these simpler usecases. 
it can maybe later be used in a complex tuning use case. but that should be separate.

File: src/usecase_regression.Rmd [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+```{r}
+## Make a task
+regr.task = makeRegrTask(data = BostonHousing, target = "medv")
+regr.task
+```
+In order to get an overview of the features type, we can print out the ``regr.task``. This shows that there are $12$ numeric and one factor variables in the data set.
+
+### Tuning
+By calling ``listLearners("regr")`` we can see which learners are available for the regression task. We picked the classical linear regression model (``regr.lm``), the [Lasso](&penalized::penalized) regression model (``regr.penalized.lasso``), [SVM](&kernlab::ksvm) with a radial basis kernel (``regr.ksvm``) and random forest from the [ranger](&ranger::ranger) package (``regr.ranger``), which we would like to compare. Before setting up a [benchmark experiment](benchmark_experiments.md) we can specify which hyperparameters are going to be tuned. The [%mlr] package provides powerful tuning algorithms. See [Tuning](tuning.md) and [Advanced Tuning](advanced_tune.md) for more details.
```
Comment (SteveBronder) [inline src/usecase_regression.Rmd]: > , which we would like to compare

I would put this at the beginning of the sentence. Maybe something like

With so many learners it is difficult to choose which one would be optimal for this specific task. As such we will choose a sample of these learners and compare their results. This analysis uses...

Comment (SteveBronder) [reply src/usecase_regression.Rmd]: > The [%mlr] package provides powerful tuning algorithms.

list a few of them, we have quite a few high profile and powerful tuning methods that we can name drop here

File: src/usecase_regression.Rmd [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+
+For each learner one hyperparameter will be tuned, i.e. the Lasso penalty parameter ``lambda1``, kernel parameter ``sigma`` for SVM model and the number of trees (``num.trees``) in the random forest model. We start with specifying a search space for each of these parameters. With [&makeTuneControlRandom] we set the tuning method to be random search. Afterwards we take $5$-fold cross validation as our [resampling strategy](resample.md) and root mean squared error (``rmse``) as optimization criterion. Finally, we make tuning wrapper for each learner.
```
Comment (SteveBronder) [inline src/usecase_regression.Rmd]: Mention to the reader that they find other parameters to tune using `getLearnerParamSet()`

Comment (SteveBronder) [reply src/usecase_regression.Rmd]: >  With [&makeTuneControlRandom] we set the tuning method to be random search. 

Random search is cool and good, but everyone else can do that too. It's simple and I like that, but maybe also do a tuning with mlrMBO or CMAES. Not necessary though

File: src/usecase_regression.Rmd [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+rdesc = makeResampleDesc("CV", iters = 5L)
+
+## Choose a performance measure
+meas = rmse
+
+## Choose a tuning method
+ctrl = makeTuneControlRandom(maxit = 100L)
+
+## Make tuning wrappers
+tuned.lasso = makeTuneWrapper("regr.penalized.lasso", rdesc, meas, ps_lasso, ctrl)
```
Comment (SteveBronder) [inline src/usecase_regression.Rmd]: for tutorials, I think it is good to explicitly name arguments when doing a function call

**Review by SteveBronder — COMMENTED** [review]
Some of these review points are heuristic, feel free to change what you like. Overall very good!

File: src/usecase_regression.Rmd (hunk 13 above)
Comment (pandeva) [reply src/usecase_regression.Rmd]: Thanks for your review! I will use CMAES. Should I discuss the input parameters of this method? Or it is enough to link the function (`makeTuneControlCMAES`) to the `cmaes` package.

Comment (SteveBronder) [reply src/usecase_regression.Rmd]: Going into too much detail could put the tutorial off track, just a link should be fine

**Comment by pat-s** [comment]
superseeded by https://github.com/mlr-org/mlr/pull/2339

**Pull request closed without merging**